### PR TITLE
attempt to hoist better errors

### PR DESCRIPTION
**Pull request by xavdid**
attempt to hoist the real issue when validation throws a generic "is not one of x,y"

before

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ = 1 =                                                                                                                        │
│     Property │ App.triggers.blah.operation                                                                                   │
│     Message  │ is not any of </BasicPollingOperationSchema>,</BasicHookOperationSchema>                                      │
│     Links    │ https://github.com/zapier/zapier-platform-schema/blob/v7.6.1/docs/build/schema.md#basicpollingoperationschema │
│              │ https://github.com/zapier/zapier-platform-schema/blob/v7.6.1/docs/build/schema.md#basichookoperationschema    │
└──────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

after

```
┌───────────────────────────────────────────────────────────────┐
│ = 1 =                                                         │
│     Property │ App.triggers.blah.operation.inputFields[0].key │
│     Message  │ does not meet minimum length of 1              │
│     Links    │                                                │
└──────────────┴────────────────────────────────────────────────┘
```

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "extends": "eslint:recommended",
   "parserOptions": {
-    "ecmaVersion": 8
+    "ecmaVersion": 2018
```
Comment (xavdid) [inline .eslintrc:4]: eslint was complaining about

```js
const {x, ...y} = someObj
```

so I bumped it. Side note, I'd love to move this to use a more standard config. [Standard.js](https://standardjs.com/) is my personal favorite, but the front-end team's, airbnb's, or others are good bets too.

File: lib/utils/makeValidator.js [hunk 2]
```diff
@@ -18,6 +21,72 @@ const makeLinks = (error, makerFunc) => {
...
+
+  err.property = completePath;
+  return err;
+};
+
+const cleanError = (validationError, path, validator) => {
+  if (ambiguousTypes.includes(validationError.name)) {
+    // TODO: make this smarter?
+    // if it's triggerSchema, check for "type"
+    // if it's request/function, check for "source" or "require". if neither are present, validate against request
```
Comment (eliangcs) [inline lib/utils/makeValidator.js]: (Non-blocker) For this TODO, do you think if it's possible to have a generic solution instead of relying on the prior knowledge we have on the app schema? I was thinking we can measure the "relevance" and find out the best match, similar to what python-jsonschema [does](https://github.com/Julian/jsonschema/blob/21838cd7727bd7c1d9a309df51cd32ebb0c78cdb/jsonschema/exceptions.py#L296).

File: lib/utils/makeValidator.js [hunk 3]
```diff
@@ -18,6 +21,72 @@ const makeLinks = (error, makerFunc) => {
...
+    // } else {
+    nextSchema = validationError.argument[0];
+    // }
+
+    // we have some schemas with anonymous subschemas (such as FieldChoices)
+    if (nextSchema.includes('[subschema')) {
+      const namedSchema = validator.schemas[validationError.schema];
+
+      if (!namedSchema) {
+        return processBaseError(validationError, path);
```
Comment (eliangcs) [inline lib/utils/makeValidator.js]: If I do this for a `perform`:

```
perform: {
  url: 'https://example.com',
  body: 123
}
```

I see this error:

```
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ = 1 =                                                                                                          │
│     Property │ App.triggers.movie.operation.perform.body                                                       │
│     Message  │ is not exactly one from [subschema 0],[subschema 1],[subschema 2],[subschema 3]                 │
│     Links    │ https://github.com/zapier/zapier-platform-schema/blob/v7.6.1/docs/build/schema.md#[subschema 0] │
│              │ https://github.com/zapier/zapier-platform-schema/blob/v7.6.1/docs/build/schema.md#[subschema 1] │
│              │ https://github.com/zapier/zapier-platform-schema/blob/v7.6.1/docs/build/schema.md#[subschema 2] │
│              │ https://github.com/zapier/zapier-platform-schema/blob/v7.6.1/docs/build/schema.md#[subschema 3] │
└──────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Is it possible to replace `[subschema x]` with a more readable text?

**Review by eliangcs — COMMENTED** [review]
The unreadability of `anyOf` error messages is something that bugs me for a long time. Thanks for fixing it! I tested locally and found some error messages are still unclear when it comes to anonymous schema. I wonder if it's fixable.

File: lib/utils/makeValidator.js (hunk 2 above)
Comment (xavdid) [reply lib/utils/makeValidator.js]: that... is very cool. It's a little weird because it looks like they just de-prioritize the `anyOf` errors. In that link, `STRONG_MATCHES` is an empty set, so i'm a little confused at how much value there is there. 

And yeah, i was going back and forth here. On the one hand, if we _only_ fish on polling/hook operation schema, we're already seeing a big improvement. doing more is even better, but i'd happily ship with just that.

Comment (xavdid) [reply lib/utils/makeValidator.js]: One idea I had was to recurse down each option (in this case `pollingOperation` and `hookOperation`) and then surface whichever path had fewer errors. Basically, the way we do it now, we could send someone down the wrong path if they do have a very malformed string request or something. Very possible I'm overthinking it though. 

My other thought is fix polling vs hook explicitly (since that's the most confusing one - it's usually an issue with an input field or something and it doesn't actually matter which of the two we validate against) and then putting a line in for all other cases of `anyOf` that says "Please file an issue if you see this so we can better cover this case". And then we can build out this section as needed.

what do you think?

File: lib/utils/makeValidator.js (hunk 3 above)
Comment (xavdid) [reply lib/utils/makeValidator.js]: maybe! i noticed the same thing. I think it's a problem of going one level too far down. I'll jump back in, but my guess is that the error structure is something like

```js
{
  argument: ['subschema 0', 'subschema 1', ...],
  schema: [{type: 'number'}, {type: 'object'}... ]
  ...
}
```

so we could pull those types out. I agree with the above that we should be careful doing this too coupled to the structure of the schema itself, but some of these edge cases are pretty wild.

**Comment by xavdid** [comment]
Ok @eliangcs I've cleaned it up and done what I can with the links and errors. I don't think it's perfect, but it's a step in the right direction and we'll be able to iterate on it as needed.

**Comment by xavdid** [comment]
Oh and the python implementation was fine, but changed the structure of the errors enough that it wasn't going to be worth the time to redo the cli code to accomodate it. Plus, `best_error` only returned a single item from the entire validation while there could be multiples across an entire app.

File: lib/utils/makeValidator.js [hunk 4]
```diff
@@ -3,21 +3,116 @@
...
+const removeFirstAndLastChar = s => s.slice(1, -1);
+// always return a string
+const makePath = (path, newSegment) =>
+  (path ? [path, newSegment].join('.') : newSegment) || '';
+
+const processBaseError = (err, path) => {
+  const completePath = makePath(path, err.property).replace(
+    /\.instance\./g,
+    '.'
+  );
```
Comment (eliangcs) [inline lib/utils/makeValidator.js]: I see there are some cases where the path ends with `.instance`. Maybe we can change it to something like:
```js
  const completePath = _.trimEnd(
    makePath(path, err.property).replace(/\.instance($|\.)/g, '.'),
    '.'
  );
```

File: lib/utils/makeValidator.js [hunk 5]
```diff
@@ -3,21 +3,116 @@
...
+      }
+
+      const res = validator.validate(validationError.instance, nextSchema);
+      if (!res.errors.length) {
+        // This is a case that I can no longer reproduce, but it caused really weird behavior so I'm going to leave this as a canary for now
+        // in theory this should never happen - all schemas at this point should have errors
+        throw new Error(
+          'If you see this, please file an issue at https://github.com/zapier/zapier-platform-cli/issues'
+        );
+      }
```
Comment (eliangcs) [inline lib/utils/makeValidator.js:99]: Throwing a hard error here blocks the dev from continuing their work, which is worse than the original behavior IMO. Maybe we want to take a milder approach here. Can we print a warning message and keep validation going, so the dev can see the validation errors at least?

I found one of the reasons why `res.errors` is an empty array below is because of this jsonschema [bug](https://github.com/tdegrunt/jsonschema/issues/263), where `validationError.instance` is undefined when the value being validated is falsy. To reproduce, try using a falsy value like an empty string in `inputFields`:

```js
{
  inputFields: [''],
}
```

To work around this bug, I suggest we do this before we do `validator.validate(validationError.instance, nextSchema)`:

```js
      if (validationError.instance === undefined) {
        // Work around a jsonschema bug: When the value being validated is a
        // falsy, such as false and '', validationError.instance isn't available.
        // See https://github.com/tdegrunt/jsonschema/issues/263
        const fullPath = path.replace(/^instance\./, '') + validationError.property.replace(/^instance\./, '.');
        validationError.instance = _.get(definition, fullPath);
      }
```

Important: You need to pass `definition` to `cleanErrors` if you want to use this code.

File: test/readability.js [hunk 6]
```diff
@@ -12,13 +13,11 @@ describe('readability', () => {
...
     results.errors.should.have.length(1);
-    results.errors[0].stack.should.eql(
-      'instance.test is not exactly one from </RequestSchema>,</FunctionSchema>'
-    );
+    results.errors[0].stack.should.eql('instance is not of a type(s) object');
   });
 
   it('should have decent messages for minimum length not met', () => {
-    const results = CreateSchema.validate({
+    const results = TriggerSchema.validate({
```
Comment (eliangcs) [inline test/readability.js:21]: Any reasons why this change?

**Review by eliangcs — APPROVED** [review]
I dug into this a bit. This is truly not an easy problem. Thanks for the effort! There's a couple of edge cases I think we should fix before we merge, but I'm approving this in advance not to block you.

File: lib/utils/makeValidator.js (hunk 4 above)
Comment (xavdid) [reply lib/utils/makeValidator.js]: yeah, good call. If you've still got it, it would be great to get an input + schema that generated that so I can write a test against it

File: test/readability.js (hunk 6 above)
Comment (xavdid) [reply test/readability.js:21]: Triggerschema operation is itself an `anyOf` (polling/hook) so I wanted to make sure errors in a nested field  (like `inputData`) correctly surfaces in that situation. Sort of the base case we're trying to handle

File: lib/utils/makeValidator.js (hunk 5 above)
Comment (xavdid) [reply lib/utils/makeValidator.js:99]: ha, that's why I made [this change](https://github.com/zapier/zapier-platform-schema/pull/62/files#diff-f31cacec6592fa8d4d949b17e59b2669R49) but didn't connect the dots. Great find!

The reason I was throwing the error was that there _should_ be an error here (since validation has failed) but we have no error to throw. If we don't throw, we fail to validate but there's nothing to show, which I think is worse behavior. Given that, do you think it's appropriate to throw an error?

I do like pulling the instance out from above though, if we can.

Comment (eliangcs) [reply lib/utils/makeValidator.js:99]: I was thinking like what if they have other errors? So for example, if I do `inputFields: [false]` and another violation like a trigger description being too short, this would just show me this block of red text:



I'd have no idea what's wrong with my code if I've ever seen this. However, if we print a warning message instead of throwing an error over this, we can still see the other error:



All that is to say even if validation has failed on one branch, we can still keep running validation on the other branches. Does that make sense?

File: lib/utils/makeValidator.js (hunk 4 above)
Comment (eliangcs) [reply lib/utils/makeValidator.js]: One example is when you validate a primitive value against `FieldSchema`. Like `inputFields: [123]`.

File: lib/utils/makeValidator.js (hunk 5 above)
Comment (xavdid) [reply lib/utils/makeValidator.js:99]: ah, sure. I guess in my head we could return an error instead of throwing it because we don't want to actually halt execution. good call!
